[Android] Add callback to host-runtime contract for getting assembly data

grendello opened the pull request:
Add a new callback to `host_runtime_contract` that is called default assembly resolution:
```c++
// Probe the host for `path`. Sets pointer to data start and its size, if found.
// Returns true if found, false otherwise. If false, out parameter values are ignored by the runtime.
bool(HOST_CONTRACT_CALLTYPE* external_assembly_probe)(
    const char* path,
    /*out*/ void** data
    /*out*/ int64_t* size);
```

Contributes to https://github.com/dotnet/runtime/issues/112706

This also makes fatal errors that the runtime prints to `stderr` get printed to logcat on Android.

This is a set of minimal changes required to make https://github.com/dotnet/android/pull/9572 (the `.NET for Android` CoreCLR host) work with standard .NET for Android applications.

grendello committed:
Beginnings of native Android build

grendello committed:
Beginnings of native Android build

grendello committed:
Beginnings of native Android build

steveisok committed:
Bump min API level to 29 (proper TLS support) and link the right system.native library for the singlefilehost

steveisok committed:
Fix host/target confusion in R2R tool publish / Add Android target to R2R that acts like linux

ivanpovazan committed:
Cleanup

steveisok committed:
Revert changes to R2R that alias Android to linux and put it in a few msbuild places instead.

steveisok committed:
Undo host awareness in crossgen_publish as building it is about the target

steveisok committed:
Make sure android runtime pack is generated and crossgen_inbuild uses host lib prefix/suffix on cross builds

steveisok committed:
Condition min api level for coreclr and not.

steveisok committed:
Add FEATURE_NATIVEAOT_ONLY

steveisok committed:
Revert FEATURE_NATIVEAOT_ONLY

steveisok committed:
Set PrimaryRuntimeFlavor on mobile official build and don't try to generate android coreclr runtime packs during a mono android build

steveisok committed:
Override API level in coreclr runtime.proj and flow runtimeFlavor into build-commons.sh

steveisok committed:
Disable ThreadStatic optimizations until emulated TLS support is added

steveisok committed:
Using API level 21 should be good now

kotlarmilos committed:
Set runtime feature flag in cmake config

steveisok committed:
Don't get in the way of building coreclr+mono runtimes together. Works on arm64 right now

steveisok committed:
Allow cross build on osx when targeting arm32, fix bad arm32 linker arg, and exclude asm routines on x64 build

steveisok committed:
Revert CLR_CMAKE_RUNTIME_CORECLR and enable event pipe TCP stuff

steveisok committed:
Exclude arm and x86 from generating packages / Always set CrossBuild=true when dealing with android

grendello committed:
Initial Android CoreCLR initialization function implementation

Not tested at run time yet (require some changes in the XA runtime
first)

grendello committed:
Preparing bundle mode for Android

grendello committed:
Tracking why System.Private.CoreLib.dll load fails

grendello committed:
Don't try to mmap assembly image on Android

grendello committed:
Tracking abort() calls

grendello committed:
For now

grendello committed:
Revert "For now"

This reverts commit 1f6ddc2714f71d0bed6300158ad121cfcb82cb8b.

grendello committed:
Logging to Android logcat

grendello committed:
Write to logcat instead of stderr

grendello committed:
This should work...

AaronRobinsonMSFT commented on src/coreclr/dlls/mscoree/exports.cpp:
```diff
@@ -130,6 +147,31 @@ static void InitializeStartupFlags(STARTUP_FLAGS* startupFlagsRef)
 }
 
+#if defined(TARGET_ANDROID)
+static void ConvertConfigPropertiesToUnicode(
```
Why is this version needed?

grendello replied:
The Android version is different enough that I thought it would make it clearer to put it in a separate version of the function, and keep the name for familiarity.

AaronRobinsonMSFT commented on src/native/corehost/host_runtime_contract.h:
```diff
@@ -38,6 +38,12 @@ struct host_runtime_contract
         size_t value_buffer_size,
         void* contract_context);
 
+#if defined(TARGET_ANDROID)
+    bool(HOST_CONTRACT_CALLTYPE* android_bundle_probe)(
+        const char* path,
+        void **data_start,
+        int64_t* size);
+#endif
```
Why can't we use the `bundle_probe`? Is the `data_start` not an offset into the application?

grendello replied:
It's essentially because assemblies can be in different places and in different storage modes:

  1. Assemblies are stored in an indexed binary blob called an "assembly store".
     This blob is placed inside an ELF shared library (an Android system requirement for machine-specific code) which, in turn, can be found
     either in one of the APK archives comprising the application (there can be one or more of them, depending on application deployment mode) or
     on the filesystem somewhere on device.
  2. Assemblies in the store can, but don't have to be, be compressed by us using the LZ4 compression.
     If they are compressed, we prepare for lazy extraction at runtime and pre-allocate appropriately sized buffers in a native ELF shared library
     built together with the application, and decompress the data into the buffers before handing pointer to the buffer out to the runtime.
     If assemblies aren't compressed, we return a pointer to read-only data which is an offset into the mmapped assembly store (which, again, can
     live either on the filesystem or in an APK ZIP archive)
  3. In Debug builds, assemblies can (but don't have to) be stored on the filesystem directly.
     If FastDev (a mode where assemblies are not linked and synchronized from the development PC to the device) is enabled, individual `.dll` files
     are placed in a dynamically selected location on the filesystem and read from there instead of from the assembly stores.

elinor-fung replied:
Yeah, it is definitely a bit different from `bundle_probe`. I think we should add it as just a new function pointer for the contract though, rather than being Android-only. I had added callback with the same signature as part of my local investigation: https://github.com/dotnet/runtime/issues/112706.

AaronRobinsonMSFT commented on src/native/corehost/host_runtime_contract.h:
```diff
@@ -53,4 +59,17 @@ struct host_runtime_contract
         const char* entry_point_name);
 };
 
+#if defined(TARGET_ANDROID)
+struct host_configuration_property
+{
+    const char16_t* name;
+    const char16_t* value;
+};
+
+struct host_configuration_properties
+{
+    size_t nitems;
+    const host_configuration_property* data;
+};
+#endif
```
We have a few collections for properties. Why is this one needed?

/cc @elinor-fung

grendello replied:
It was part of the original contract, but the properties were passed as 3 parameters to `coreclr_initialize`.  Passing them as one pointer to a structure makes it use less registers on call and makes it easier to pass a `NULL` pointer to indicate there are no properties.  Also, `coreclr_initialize` expects UTF-8 properties, this structure uses Unicode.

AaronRobinsonMSFT replied:
> Passing them as one pointer to a structure makes it use less registers on call and makes it easier to pass a NULL pointer to indicate there are no properties. Also, coreclr_initialize expects UTF-8 properties, this structure uses Unicode.

I really don't think we should be optimizing at this point. All of this is in service of getting things going and then we can work to create/design entry points that work. I get this is quick and dirty, but the quick and dirty should be on the consuming side when possible, not the runtime. There has been much talk about this API and how it is less than ideal, I don't think this PR is where we need to hack in a new one.

grendello replied:
This part can be removed then, however calling `coreclr_initialize` as it is now, with pointer-to-string-to-pointer conversions is not an option IMO, so we need to work around that aspect somehow. If we can modify `coreclr_initialize` to not expect the pointer-as-string abstraction, then sure, we can use it instead of the new API.

AaronRobinsonMSFT commented on src/coreclr/vm/util.cpp:
```diff
@@ -179,8 +182,13 @@ void PrintToStdErrA(const char *pszString) {
 
     HANDLE  Handle = GetStdHandle(STD_ERROR_HANDLE);
 
+#if defined(TARGET_ANDROID)
+    // TODO: make priority configurable?
+    __android_log_write(ANDROID_LOG_INFO, "CoreCLR", pszString);
+#else
     size_t len = strlen(pszString);
     NPrintToHandleA(Handle, pszString, len);
+#endif
 }
```
This seems like it should be pushed into the `NPrintToHandleA` function. All calls to `NPrintToHandleA` seem like they should be pushed to `__android_log_write`.

grendello replied:
Logging changes most probably require a thorough review from all the interested parties, what's in this PR is merely a set of quick hacks to put something in the logcat buffer.

AaronRobinsonMSFT commented on src/coreclr/vm/peimage.inl:
```diff
@@ -98,8 +111,11 @@ inline const SString &PEImage::GetModuleFileNameHintForDAC()
     WRAPPER_NO_CONTRACT;
-
+#if defined(TARGET_ANDROID)
+    return false;
```
```suggestion
    return FALSE;
```

AaronRobinsonMSFT commented on src/coreclr/vm/peimage.inl:
```diff
@@ -32,12 +32,25 @@ inline const SString &PEImage::GetPath()
 }
 
+#if defined(TARGET_ANDROID)
+inline const HANDLE PEImage::AndroidGetDataStart()
```
Can't this be `GetFileHandle()`?

grendello replied:
I'd rather keep this name, as the pointer isn't really a file handle - it's a pointer to a memory area which may, but doesn't have to, correspond to a file.

steveisok replied:
@AaronRobinsonMSFT is `FileHandle` what we usually call such a thing in runtime? I think it would make sense to use whatever is common in runtime.

jkotas replied:
`HANDLE` used for things that you dispose by calling `CloseHandle`. This is not a handle. This is a pointer.

The type should be `void*` if the external mapping is R/W; or `const void* if it is R/O.

The name can be DataStart or just Data. However, it is a security best practice to always pass a buffer size together with the pointer the buffer, so I would make the function return the size of the payload too.

grendello replied:
The size is stored alongside the pointer in `m_bundleFileLocation`, and this function is a variation on what's already done in `Bundle` mode.  In the current implementation, the pointer to data is implied to be the start of the bundle file, with offset returned by `PEImage::GetOffset()` and size returned by `PEImage::GetSize()`. The change for this function to return both the pointer and the size is, I believe, another item for a follow-up since the change would be much larger in scope than the current PR. Also, I don't know the code good enough to be comfortable in making this change at this point.

Please keep in mind that this PR is meant to be merely a beginning of a longer process, so changes like what you propose can and should be made in follow-up PRs.

AaronRobinsonMSFT commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -869,6 +876,9 @@ HRESULT PEImage::TryOpenFile(bool takeLock)
 
     ErrorModeHolder mode{};
+#if defined(TARGET_ANDROID)
+    m_hFile = AndroidGetDataStart ();
```
```suggestion
    m_hFile = AndroidGetDataStart();
```

AaronRobinsonMSFT commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -881,7 +891,7 @@ HRESULT PEImage::TryOpenFile(bool takeLock)
                           FILE_ATTRIBUTE_NORMAL,
                           NULL);
-
+#endif
```
```suggestion
#endif // TARGET_ANDROID
```

AaronRobinsonMSFT commented on src/coreclr/vm/common.h:
```diff
@@ -397,7 +397,7 @@ extern DummyGlobalContract ___contract;
 
 void LogErrorToHost(const char* format, ...);
-
+#if defined(TARGET_ANDROID)
+void LogDebugToHost(const char* format, ...);
+#endif
```
@elinor-fung What is the best way to express debug logging to the host?

elinor-fung replied:
We only have an error writing callback (`coreclr_set_error_writer`) currently. Do we actually need debug logging to the host? The error logging to host is largely for runtime initialization issues, so the host can report them to the user. For debug logging once the runtime is actually going, I'd expect we could update the runtime logging mechanisms such that they use Android logging APIs instead.

AaronRobinsonMSFT commented on src/coreclr/hosts/inc/coreclrhost.h:
```diff
@@ -22,13 +22,26 @@
     typedef int (CORECLR_CALLING_CONVENTION *function##_ptr)(__VA_ARGS__)
 
+#if defined(TARGET_ANDROID)
+CORECLR_HOSTING_API(android_coreclr_initialize,
+            const char* appName,
+            const char16_t* appDomainFriendlyName,
+            host_runtime_contract* hostContract,
+            const host_configuration_properties* properties,
+            void **hostHandle,
+            unsigned int* domainId);
```
This new signature looks unnecessary. Why can't we re-use `coreclr_initialize`?

grendello replied:
`coreclr_initialize` requires that we pass pointers encoded as strings, which are then parsed back to pointers. This is both slow and not really necessary when we link `libcoreclr.so` directly - we should just pass the raw pointer. The pointer-to-string-to-pointer conversion can also break tagged pointers.  We also probably don't need `appName` and `appDomainFriendlyName`, but I kept them because I wasn't sure if that is really the case. This also allows us to use the new configuration properties encoded as Unicode. Using `coreclr_initialize` would require modification to implement these changes and, thus, break compatibility with already existing hosts that use `coreclr_initialize`

AaronRobinsonMSFT commented on src/coreclr/dlls/mscoree/mscorwks_unixexports.src:
```diff
@@ -1,3 +1,6 @@
+; Android hosting API
+android_coreclr_initialize
```
We should not be adding new public APIs to `coreclr.dll`.

grendello replied:
I agree, but I didn't want to make bigger changes and create a separate script for Android.

AaronRobinsonMSFT commented on src/coreclr/dlls/mscoree/exports.cpp:
```diff
@@ -509,3 +652,13 @@ void LogErrorToHost(const char* format, ...)
     }
 }
+
+#if defined(TARGET_ANDROID)
+void LogDebugToHost(const char* format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    __android_log_vprint(ANDROID_LOG_DEBUG, ANDROID_LOGCAT_TAG, format, args);
+    va_end(args);
+}
+#endif
```
I would follow the `LogErrorToHost` pattern here. @elinor-fung Can provide guidance on if we should be creating a new `g_errorWriter` option for debug. Perhaps this is a NOP on non-Android for now.

elinor-fung replied:
I don't actually see anything using `LogDebugToHost`?

grendello committed:
Update src/coreclr/vm/peimage.inl

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

grendello committed:
Update src/coreclr/vm/peimage.cpp

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

lewing commented:
/azp run runtime

AaronRobinsonMSFT commented:
@grendello I think a bunch of these changes make sense to me, for example, a bunch of the bundling logic. My pushback would be either we are doing something quick and dirty or we are chasing optimizations, not both. For example `coreclr_initialize` seems to be just fine, but we don't want to have the other side convert to UTF-8, why not? Would that make composition harder in some way? That is the existing API and we should keep that for now to get things up.

I would change the android side to conform, where possible, to existing APIs for now and put a bunch of the cost there.

jkotas commented:
This change was incorrectly merged with main and a lot of unrelated commits from main are showing up as part of the delta. Could you please rebase against main and force push so that the delta is reviewable?

grendello commented:
> This change was incorrectly merged with main and a lot of unrelated commits from main are showing up as part of the delta. Could you please rebase against main and force push so that the delta is reviewable?

I use `git merge`, which is the correct way of tracking changes, because it doesn't rewrite history like `git rebase` does, but sure, I can rebase.

grendello commented:
> @grendello I think a bunch of these changes make sense to me, for example, a bunch of the bundling logic. My pushback would be either we are doing something quick and dirty or we are chasing optimizations, not both. For example `coreclr_initialize` seems to be just fine, but we don't want to have the other side convert to UTF-8, why not? Would that make composition harder in some way? That is the existing API and we should keep that for now to get things up.

Using UTF-8 strings on the Android side is easy and bears no costs for us, I can make the change without any effort. This is not the only issue with `coreclr_initialize`, though, and I'm going to push back against using the pointer-to-string-to-pointer conversion (as I did years ago when the same approach was suggested for MonoVM). 

> 
> I would change the android side to conform, where possible, to existing APIs for now and put a bunch of the cost there.

:+1:

grendello committed:
Post-rebase adjustments

grendello committed:
[WIP] Remove android_coreclr_initialize and modify the existing one

Signature is altered on Android, by adding a pointer to the host contract
structure.

The host contract structure also gains a new pointer, `external_assembly_probe`
(proposed in https://github.com/dotnet/runtime/issues/112706) which is used, if
present, when `bundle_probe` is `nullptr` or if it returns `false` when called.

Either `bundle_probe` or `external_assembly_probe` must be set in the contract

jkotas commented:
> I use git merge, which is the correct way of tracking changes, because it doesn't rewrite history like git rebase does, but sure, I can rebase.

Yes, `git merge` is the correct way. I am not sure what you have done, but there was an operator error somewhere along the way. Maybe your local main diverged somehow?

The git notification for properly executed `git merge` looks like this - it lists the merge commit and not the actual commits included in the merge:
<img width="353" alt="image" src="https://github.com/user-attachments/assets/daefd573-db2f-4cc8-9e5a-b1725f575498" />

However, git notification for that merge that you have done looked like this - it did not list the merge commit, but it listed the actual commits included in the merge:
<img width="545" alt="image" src="https://github.com/user-attachments/assets/7a53e9b2-f761-4414-9b0d-02e2886ff275" />

jkotas commented on src/native/corehost/host_runtime_contract.h:
```diff
@@ -53,4 +59,17 @@ struct host_runtime_contract
         const char* entry_point_name);
 };
 
+#if defined(TARGET_ANDROID)
+struct host_configuration_property
+{
+    const char16_t* name;
+    const char16_t* value;
+};
+
+struct host_configuration_properties
+{
+    size_t nitems;
+    const host_configuration_property* data;
+};
+#endif
```
```suggestion
```
Not needed anymore?

steveisok replied:
Not now - I removed them. Note, anything defined here needs to be C friendly since it's shared w/ mono.

steveisok committed:
Cleanup - build failed locally for arm64. Removed unused blocks

grendello commented:
> > I use git merge, which is the correct way of tracking changes, because it doesn't rewrite history like git rebase does, but sure, I can rebase.
> 
> Yes, `git merge` is the correct way. I am not sure what you have done, but there was an operator error somewhere along the way. Maybe your local main diverged somehow?
> 
> The git notification for properly executed `git merge` looks like this - it lists the merge commit and not the actual commits included in the merge:

No idea what might have happened, I always used just `git merge main` with this branch.

steveisok committed:
Remove host_runtime_contract from coreclr_initialize. Current expectation is that it needs to be encoded as a base64 string

jkotas commented on src/coreclr/hosts/inc/coreclrhost.h:
```diff
@@ -22,6 +22,10 @@
 #define CORECLR_HOSTING_API_LINKAGE
 #endif
 
+#if defined(TARGET_ANDROID)
+#include "corehost/host_runtime_contract.h"
+#endif
+
```
```suggestion
```

jkotas commented on src/coreclr/inc/bundle.h:
```diff
@@ -18,29 +18,35 @@ class Bundle;
 
         Size = 0;
-        Offset = 0; 
+        DataStart = INVALID_HANDLE_VALUE;
```
```suggestion
        DataStart = nullptr;
```
`INVALID_HANDLE_VALUE` is meant for HANDLEs. DataStart is a pointer. It is not a good idea to mix HANDLEs and pointers.

jkotas commented on src/coreclr/vm/bundle.cpp:
```diff
@@ -30,14 +30,29 @@ const SString &BundleFileLocation::Path() const
     m_probe = probe;
+    m_externalAssemblyProbe = externalAssemblyProbe;
+#if defined(TARGET_ANDROID)
+    m_appName.SetUTF8(bundlePath);
```
Can we use the existing `m_path` to store the `bundlePath` for Android?

grendello replied:
`bundlePath` on Android isn't a path at all, it's the package (app) name.  Storing the value in `m_path` would be misleading.

steveisok replied:
I think `m_path` with a comment about why it could be misleading would achieve the same result.

jkotas replied:
> bundlePath on Android isn't a path at all, it's the package (app) name. Storing the value in m_path would be misleading.

It is about as misleading as passing it in argument named `bundlePath`.

jkotas commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -725,6 +731,11 @@ FlatImageLayout::FlatImageLayout(PEImage* pOwner)
         }
+#else // !TARGET_ANDROID
+        m_FileMap.Assign(mapBegin);
+        m_FileView.Assign(mapBegin);
```
m_FileMap and m_FileView have destructors that will try to free what they are pointing to as a file-mapping / as HANDLE.

Assigning a pointer that's not a file-mapping / a HANDLE is going to lead to inscrutable crashes when the destructor runs.

grendello replied:
This is done this way to make the PR work, I think the correct way should be implemented in the follow up. Android applications don't quit unless they are killed, so **for now**, I think it's acceptable.

jkotas replied:
I am not comfortable with merging code to main that's obviously wrong. (I am ok with merging code that's not finished.)

grendello committed:
Update src/coreclr/hosts/inc/coreclrhost.h

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

grendello committed:
Update src/coreclr/inc/bundle.h

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

grendello committed:
Update src/coreclr/hosts/inc/coreclrhost.h

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

grendello committed:
Address feedback

vitek-karas commented on src/coreclr/vm/bundle.cpp:
```diff
@@ -80,11 +102,27 @@ BundleFileLocation Bundle::Probe(const SString& path, bool pathIsBundleRelative)
+
+    if (m_probe != nullptr)
+    {
+        assemblyFound = m_probe(utf8Path, &loc.Offset, &fileSize, &compressedSize);
```
This is a slight change in the existing behavior. Currently the existing probe is defined as returning bool, but the code ignored it, now it doesn't. Given that we don't really expose this probe as an external API it's probably OK, but we should make it a concious decision.

steveisok committed:
Merge remote-tracking branch 'upstream/main' into dev/grendel/android-clr-host-local

steveisok committed:
Remove some logging related things since Elinor's PR landed

elinor-fung committed:
Fix HostRuntimeContract tests

elinor-fung commented on src/native/corehost/host_runtime_contract.h:
```diff
@@ -51,6 +51,12 @@ struct host_runtime_contract
         const char* entry_point_name);
+
+    // Probe the host for `path`. Sets pointer to data start and its size, if found.
+    bool(HOST_CONTRACT_CALLTYPE* external_assembly_probe)(
```
@grendello I moved this to the end of the struct to keep the existing order. I don't think this should require updating your changes in dotnet/android, since they use designators.

elinor-fung committed:
AndroidGetDataStart -> GetData

elinor-fung committed:
Don't try to manage external assembly not mapped by the runtime

elinor-fung committed:
Remove newline-only changes

jkotas commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -616,17 +616,26 @@ FlatImageLayout::FlatImageLayout(PEImage* pOwner)
 #endif // LOGGING
 
+// TODO: Any image from external_assembly_probe should go down this path, not just on Android
+#if defined(TARGET_ANDROID)
```
I believe that the existing Android tests in dotnet/runtime expect (unbundled) files on disk. Do we need to turn these ifdefs into runtime conditions to keep these tests working?

elinor-fung replied:
Switched to check for non-file image with non-null data.

elinor-fung committed:
Explicitly check / document checking return value of probes

elinor-fung committed:
Check for non-file data instead of assuming that is the case on Android

jkotas commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -616,17 +616,31 @@ FlatImageLayout::FlatImageLayout(PEImage* pOwner)
     LOG((LF_LOADER, LL_INFO100, "PEImage: Opening flat %s\n", ownerPath.GetUTF8()));
 #endif // LOGGING
 
+    if (!pOwner->IsFile())
+    {
```
```suggestion
```
This should not be needed.

jkotas commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -869,6 +876,9 @@ HRESULT PEImage::TryOpenFile(bool takeLock)
         return S_OK;
 
     ErrorModeHolder mode{};
+#if defined(TARGET_ANDROID)
+    m_hFile = GetData(nullptr);
+#else
```
```suggestion
```
If the `GetFileHandle` is still called after the other changes. fix the caller instead of this hack?

jkotas commented on src/coreclr/vm/peimage.h:
```diff
@@ -133,11 +133,16 @@ class PEImage final
 
     BOOL IsFile();
     BOOL IsInBundle() const;
+    void* GetData(INT64* size);
```
```suggestion
    void* GetExternalData(INT64* size);
```
To make it easier to see the connection with `external_assembly_probe`?

elinor-fung committed:
Remove unnecessary check

elinor-fung committed:
Don't set file handle for non-file data

elinor-fung committed:
Check for null data instead of always returns FALSE for PEImage::IsFile on Android

elinor-fung committed:
GetData -> GetExternalData

elinor-fung committed:
Remove AndroidGetAppName

elinor-fung committed:
Remove BundleFileLocation::AppName

elinor-fung commented on src/coreclr/vm/bundle.cpp:
```diff
@@ -64,6 +65,13 @@ BundleFileLocation Bundle::Probe(const SString& path, bool pathIsBundleRelative)
     LPCSTR utf8Path(pathBuffer.GetUTF8());
 
+#if defined(TARGET_ANDROID)
+    // On Android we always strip away the assembly path, if any
```
@grendello Thoughts on doing this stripping on the Android host side instead? For the runtime, I think it would make more sense to just pass the raw path to `external_assembly_probe`.

grendello replied:
Sure :+1:

AaronRobinsonMSFT commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -383,6 +383,7 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_LogToConsole, W("LogToConsole"), 0 , "Writes t
 CONFIG_DWORD_INFO(INTERNAL_LogToDebugger, W("LogToDebugger"), 0 , "Writes the CLR log to debugger (OutputDebugStringA).")
 CONFIG_DWORD_INFO(INTERNAL_LogToFile, W("LogToFile"), 0 , "Writes the CLR log to a file.")
 CONFIG_DWORD_INFO(INTERNAL_LogWithPid, W("LogWithPid"), FALSE, "Appends pid to filename for the CLR log.")
+CONFIG_DWORD_INFO(INTERNAL_LogToAndroid, W("LogToAndroid"), 0 , "Writes the CLR log to android logcat")
```
I don't think this is needed. Given the conversations about Android, we are really printing to the "console". I'm removing this and Android scenarios should reuse `LogToConsole`.

AaronRobinsonMSFT committed:
Remove LogToAndroid. Scenario should now use LogToConsole.
Minor nits.

elinor-fung commented on src/coreclr/utilcode/log.cpp:
```diff
@@ -372,10 +375,16 @@ VOID LogSpewAlwaysValist(const char *fmt, va_list args)
+#else
+        // TODO: priority should be configurable here (best, passed via a parameter)
+        //       likewise for the tag
+        __android_log_write(ANDROID_LOG_INFO, "CoreCLR-log", pBuffer);
```
```suggestion
        __android_log_write(ANDROID_LOG_INFO, MAIN_CLR_MODULE_NAME_A, pBuffer);
```

elinor-fung commented on src/coreclr/vm/util.cpp:
```diff
@@ -177,10 +180,14 @@ void PrintToStdErrA(const char *pszString) {
-
+#if defined(TARGET_ANDROID)
+    // TODO: make priority configurable?
+    __android_log_write(ANDROID_LOG_INFO, "CoreCLR", pszString);
```
```suggestion
    __android_log_write(ANDROID_LOG_INFO, MAIN_CLR_MODULE_NAME_A, pszString);
```

elinor-fung commented on src/coreclr/pal/src/thread/process.cpp:
```diff
@@ -2545,6 +2549,10 @@ PROCAbort(int signal, siginfo_t* siginfo)
     SEHCleanupSignals(false /* isChildProcess */);
 
+#if defined(TARGET_ANDROID)
+    __android_log_print (ANDROID_LOG_INFO, "CoreCLR", "Aborting in %s:%u", __FILE_NAME__, __LINE__);
```
```suggestion
    __android_log_print (ANDROID_LOG_FATAL, MAIN_CLR_MODULE_NAME_A, "Aborting in %s:%u", __FILE_NAME__, __LINE__);
```

elinor-fung commented on src/coreclr/vm/util.cpp:
```diff
@@ -177,10 +180,14 @@ void PrintToStdErrA(const char *pszString) {
-    HANDLE  Handle = GetStdHandle(STD_ERROR_HANDLE);
-
+#if defined(TARGET_ANDROID)
+    // TODO: make priority configurable?
```
Maybe it could just be `ANDROID_LOG_ERROR`?

AaronRobinsonMSFT replied:
Yes.

AaronRobinsonMSFT committed:
Feedback

AaronRobinsonMSFT committed:
Fix the build

jkotas commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -850,7 +850,10 @@ HANDLE PEImage::GetFileHandle()
     {
 #if !defined(DACCESS_COMPILE)
-        EEFileLoadException::Throw(GetPathToLoad(), hr);
+        EEFileLoadException::Throw(
+            GetPathToLoad(),
+            hr
+        );
```
```suggestion
        EEFileLoadException::Throw(GetPathToLoad(), hr);
```
This does not need to on multiple lines now that there is no ifdef

AaronRobinsonMSFT committed:
Update src/coreclr/vm/peimage.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jkotas commented on src/coreclr/vm/util.cpp:
```diff
@@ -177,10 +180,14 @@ void PrintToStdErrA(const char *pszString) {
-    HANDLE  Handle = GetStdHandle(STD_ERROR_HANDLE);
-
+#if defined(TARGET_ANDROID)
+    // TODO: make priority configurable?
```
This is used to print fatal errors right before the process aborts. This should be something that gets always printed to the log. I do not think there is any situation where one would not want to see it in the log.

AaronRobinsonMSFT replied:
So what you're saying is we do both, right?

AaronRobinsonMSFT replied:
Actually the ifdef'd out branch is printing to stderr so I'm not sure what is desired here.

jkotas replied:
I was commenting on the priority TODO. It should be hardcoded to highest possible priority.

stdout and stderr are useless on Android.

elinor-fung commented on src/coreclr/vm/util.cpp:
```diff
@@ -177,10 +180,14 @@ void PrintToStdErrA(const char *pszString) {
-    HANDLE  Handle = GetStdHandle(STD_ERROR_HANDLE);
-
+#if defined(TARGET_ANDROID)
+    // TODO: make priority configurable?
+    __android_log_write(ANDROID_LOG_ERROR, MAIN_CLR_MODULE_NAME_A, pszString);
```
```suggestion
    __android_log_write(ANDROID_LOG_FATAL, MAIN_CLR_MODULE_NAME_A, pszString);
```

elinor-fung committed:
Update src/coreclr/vm/util.cpp

am11 commented on eng/native/gen-buildsys.sh:
```diff
@@ -106,6 +106,7 @@ if [[ "$host_arch" == "wasm" ]]; then
     fi
 fi
 
+echo "$cmake_command --no-warn-unused-cli -G \"$generator\" \"-DCMAKE_BUILD_TYPE=$buildtype\" \"-DCMAKE_INSTALL_PREFIX=$__CMakeBinDir\" $cmake_extra_defines $__UnprocessedCMakeArgs \"${cmake_extra_defines_wasm[@]}\" -S \"$1\" -B \"$2\""
```
We are using this pattern to deduplicate command echo-then-eval: https://github.com/dotnet/runtime/blob/0205063ee6f87c4ac3bc22a48b71d930a03cbe5b/eng/native/build-commons.sh#L174-L176 keep it consistent?

steveisok committed:
Remove stripping away path. Dedup script echo

ivanpovazan commented on src/coreclr/dlls/mscordbi/CMakeLists.txt:
```diff
@@ -106,6 +106,12 @@ elseif(CLR_CMAKE_HOST_UNIX)
         )
     endif(CLR_CMAKE_HOST_HAIKU)
 
+    if(CLR_CMAKE_TARGET_ANDROID)
+        list(APPEND COREDBI_LIBRARIES
+            log
+        )
+    endif(CLR_CMAKE_TARGET_ANDROID)
+
```
@grendello @janvorli @lateralusX 
I suggest we move the logging bits into a separate PR which would contribute to: https://github.com/dotnet/runtime/issues/112865 I just created

steveisok committed:
Remove logging - ivan will follow up in a separate PR

jkotas committed:
Merge branch 'main' into dev/grendel/android-clr-host-local

ivanpovazan commented on src/coreclr/hosts/inc/coreclrhost.h:
As we have changes in hosting header files (this one and `src/native/corehost/host_runtime_contract.h` below), I think we could consider shipping hosting headers as part of the runtime pack?

@elinor-fung @jkotas what do you think?

steveisok replied:
Can we ship headers and still consider them not guaranteed to be stable? That's what mono does with certain parts of its hosting API.

ivanpovazan replied:
In addition to what Steve said:

Currently, on dotnet/macios side the build downloads the `coreclrhost.h` header file to build the CoreCLR host:
https://github.com/dotnet/macios/blob/30cce24d4c18d149879fe8afa64331491535ea0b/runtime/Makefile#L38
so with every change to the file, the download URL needs to be updated to have consistent behavior. This is difficult to maintain as the version of the file needs to be aligned with the runtime pack version including the actual implementation changes. 
As we are now bringing CoreCLR support for more platforms and as we are in the development stage syncing changes between the runtime and embedders/hosts, I was curious if we would want to ship the required header files as part of the runtime pack (like mono does), to avoid manual downloads or potential mismatches.

steveisok replied:
There's always the option of sharing a transport package if we don't want to ship it.

elinor-fung replied:
I don't think we want the internal coreclr interface as part of the shipped runtime pack. We do include public hosting headers ([coreclr_delegates.h](https://github.com/dotnet/runtime/blob/main/src/native/corehost/coreclr_delegates.h), [hostfxr.h](https://github.com/dotnet/runtime/blob/main/src/native/corehost/hostfxr.h), [nethost.h](https://github.com/dotnet/runtime/blob/main/src/native/corehost/nethost.h)) in the host pack.

Transport package seems reasonable if we need a better way for other repos to get at internals.

steveisok committed:
Remove redundant target_link_libraries

jkotas committed:
Merge branch 'main' into dev/grendel/android-clr-host-local

jkotas committed:
Merge branch 'main' into dev/grendel/android-clr-host-local

jkotas reviewed (approved):
The change LGTM once we get a successful CI run.

janvorli commented on src/coreclr/vm/common.h:
```diff
@@ -397,7 +397,4 @@ extern DummyGlobalContract ___contract;
 #endif
 
 void LogErrorToHost(const char* format, ...);
-
```
A nit - can we revert the change in this file as it is a pure whitespace change?

janvorli reviewed (approved):
LGTM, thank you!

elinor-fung committed:
Revert files with whitespace-only changes

elinor-fung commented:
Follow-ups I plan on for https://github.com/dotnet/runtime/issues/112706:
- host tests for `external_assembly_probe`
- refactoring/renaming to separate external assembly probe from bundle / bundle file location idea
  - external assembly probe doesn't have to mean bundle - plugging into the bundle probing was convenient, since it was already hooked in to be called before all the default resolution, but from the runtime side, external assembly data from the host shouldn't equate to bundle
  - `BundleFileLocation` name seems misleading now that it also represents externally provided assembly data

steveisok commented:
Thanks everyone for working together and helping to see this through!

steveisok committed:
Merge branch 'main' into dev/grendel/android-clr-host-local

steveisok committed:
Merge branch 'main' into dev/grendel/android-clr-host-local

jkoritzinsky commented on eng/native/build-commons.sh:
I think these changes may break source-build, as it effectively regresses https://github.com/dotnet/runtime/issues/35727. cc: @omajid can you take a look?

@grendello What was the motivation behind this change?

am11 replied:
This change is expanding the coverage of https://github.com/dotnet/runtime/issues/35727 to cmake configure step. How is it regressing, do you have an example?

omajid replied:
From https://github.com/dotnet/runtime/issues/35727:

> The normal way to do that is by exporting environment variables like CFLAGS and CXXFLAGS. Fedora provides a bunch of values that we can use . But this doesn't play very nice with the liberal use of -Werror in corefx. Worse, the flags leak into various configure tests . For example HAVE_IN_PKTINFO would fail due to an unrelated warning

am11 replied:
> it just seemed to be a mistake to have the exports without any effect.

Before this PR extra flags did had effect but only over `cmake --build` step, not the config step.

If something is breaking because of this, it is most likely the problem which needs to be fixed in that config.

grendello replied:
When you build with Ninja (don't know about Make), the `*FLAGS` are taken account only during the configuration phase. The resulting `build.ninja` file doesn't make use of them.

am11 replied:
Yup, I just tested it they seem to have no effect with ninja and make. Perhaps it was changed since https://github.com/dotnet/runtime/issues/35727?

sidenote: since it was unused before, I think we should make the names consistent with the other two user-facing `CLR_CC` and `CLR_CXX` env vars (which we use in init-compiler): `CLR_CFLAGS`, `CLR_CXXFLAGS`, `CLR_LDFLAGS`

omajid replied:
@am11 @grendello The distribution policies are around setting the flags at build time.

As we found in https://github.com/dotnet/runtime/issues/35727, setting the flags at configure-time can make .NET configure tests fail incorrectly, and result in cmake making the wrong decision about the system configuration. So when .NET is finally built, it's built in the wrong configuration - the build may fail or might produce a runtime that's buggy.

In Fedora in the past, we have ended up linking against the wrong variant of strerror_r because of this: https://pagure.io/dotnet-sig/dotnet-2-1/c/29673bc6dc37f205890fa1090e58d757bd262afb?branch=master

steveisok replied:
Should we revert the part of this change that impacts source build? Or are we ok with fixing outside of runtime?

omajid replied:
Thanks, @am11 and @grendello !

@steveisok said:

> Or are we ok with fixing outside of runtime?

Based on the comments and feedback, I think we should fix it outside of runtime.
